### PR TITLE
[MRG+1] Py3: port http downloaders

### PR DESCRIPTION
**Pull request by lopuhin**
This ports ``tests/test_downloader_handlers.py`` to py3 (with ftp tests skipped), and allows to run simple spider under py3. Is is a continuation of #1676, that contains all changes from there. Approach is the same. In some places, where twisted expects unicode args, we decode host from bytes to unicode, because ``scrapy.core.downloader.webclient._parse`` is used here as well, and returns all url parts as bytes.

I also filed an issue against twister here https://twistedmatrix.com/trac/ticket/8175 - this part remains skipped on py3.

~~Maybe https tunnel without noconnect is missing a test: I will check when porting other downloader tests and add if needed.~~

### CODE REVIEW

**Comment by lopuhin** [comment]
Will be also adding tests for https tunnel and response url validation

**Comment by lopuhin** [comment]
Also tried to make a test for https with HTTP11DownloadHandler, but it fails with "Invalid DNS-ID".

**Comment by lopuhin** [comment]
Checked that https path here https://github.com/scrapy/scrapy/pull/1678/files#diff-18150b1d259c93bf10bf1d4e5028d753R186 is not tested, so will need to add a test for it to finish porting this module. And ``TunnelingTCP4ClientEndpoint`` in the same file also needs test (and likely minor fixes).

**Comment by lopuhin** [comment]
Currently skipped https test ``test_timeout_download_from_spider`` fails to cleanup, not sure if is some test artifact or real issue, and how to dig it out:
```
_______________________________________________________________ Https11TestCase.test_timeout_download_from_spider ________________________________________________________________
NOTE: Incompatible Exception Representation, displaying natively:

DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<<class 'twisted.internet.tcp.Client'> to ('localhost', 58185) at 10d488750>
```

**Comment by lopuhin** [comment]
All is tested now. Although I still have no solution for this https://github.com/scrapy/scrapy/pull/1678#issuecomment-172482929

File: scrapy/http/response/__init__.py [hunk 1]
```diff
@@ -34,7 +35,7 @@ def _get_url(self):
         return self._url
 
     def _set_url(self, url):
-        if isinstance(url, str):
+        if isinstance(url, six.string_types):
```
Comment (kmike) [inline scrapy/http/response/__init__.py]: why is this change needed?

File: tests/test_downloader_handlers.py [hunk 2]
```diff
@@ -297,27 +312,30 @@ def test_download(self):
...
 
-            # download_maxsize = 50 is enough for the gzipped response
-            failure = crawler.spider.meta.get('failure')
-            self.assertTrue(failure == None)
-            reason = crawler.spider.meta['close_reason']
-            self.assertTrue(reason, 'finished')
+            if six.PY2:
+                # download_maxsize = 50 is enough for the gzipped response
+                # See issue https://twistedmatrix.com/trac/ticket/8175
+                raise unittest.SkipTest("xpayload only enabled for PY2")
```
Comment (kmike) [inline tests/test_downloader_handlers.py]: why is there code after raise?

File: scrapy/http/response/__init__.py (hunk 1 above)
Comment (lopuhin) [inline scrapy/http/response/__init__.py]: Oops, this is not needed, reverted

File: tests/test_downloader_handlers.py (hunk 2 above)
Comment (lopuhin) [inline tests/test_downloader_handlers.py]: I messed-up indentation, sorry - pushed a fix. A part of the test is skipped on py3, I think I should be able to unskip it in #1680, where compression is ported.

Comment (lopuhin) [inline tests/test_downloader_handlers.py]: Ah, no, sorry - this is even a comment about why this is skipped (a py3 bug in Twisted), I moved the comment.

File: scrapy/core/downloader/handlers/http11.py [hunk 3]
```diff
@@ -180,10 +185,11 @@ def _get_agent(self, request, timeout):
         if proxy:
             _, _, proxyHost, proxyPort, proxyParams = _parse(proxy)
             scheme = _parse(request.url)[0]
-            omitConnectTunnel = proxyParams.find('noconnect') >= 0
-            if  scheme == 'https' and not omitConnectTunnel:
+            proxyHost = to_unicode(proxyHost)
+            omitConnectTunnel = proxyParams.find(b'noconnect') >= 0
```
Comment (kmike) [inline scrapy/core/downloader/handlers/http11.py]: is it just a weird way to write `b'noconnect' in proxyParams`?

Comment (lopuhin) [inline scrapy/core/downloader/handlers/http11.py]: Yes :) Will fix

File: scrapy/http/response/__init__.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 See documentation in docs/topics/request-response.rst
 """
+import six
```
Comment (kmike) [inline scrapy/http/response/__init__.py]: unused import?

Comment (lopuhin) [inline scrapy/http/response/__init__.py]: yes, removed

**Comment by kmike** [comment]
+1 to merge it and open a separate issue to track https://github.com/scrapy/scrapy/pull/1678#issuecomment-172482929

**Comment by lopuhin** [comment]
I created the issue for this tests - #1696

File: scrapy/core/downloader/handlers/http11.py [hunk 5]
```diff
@@ -91,11 +92,15 @@ def __init__(self, reactor, host, port, proxyConf, contextFactory,
 
     def requestTunnel(self, protocol):
         """Asks the proxy to open a tunnel."""
-        tunnelReq = 'CONNECT %s:%s HTTP/1.1\r\n' % (self._tunneledHost,
-                                                  self._tunneledPort)
+        tunnelReq = (
+            b'CONNECT ' +
+            to_bytes(self._tunneledHost, encoding='ascii') + b':' +
+            to_bytes(str(self._tunneledPort)) +
+            b' HTTP/1.1\r\n')
```
Comment (eliasdorneles) [inline scrapy/core/downloader/handlers/http11.py]: it's a pity that % formatting only works for bytes in Python 3.5 :/

Comment (kmike) [inline scrapy/core/downloader/handlers/http11.py]: we can also build an unicode string and then convert it to bytes instead of converting individual parts

Comment (eliasdorneles) [inline scrapy/core/downloader/handlers/http11.py]: yea, was just thinking that -- would be cleaner.
the string concatenation reminds of Java too much. :)

**Comment by eliasdorneles** [comment]
Thanks so much @lopuhin, awesome work! :bow: :bow:

**Comment by redapple** [comment]
Merci @lopuhin !
